Caroline: headphone plug/unplug & kernel 4.16.18 support

bcarnes opened the pull request.

Commit:
Add headphone plug/unplug support for caroline

Commit:
apply headphone jack detection via acpid immediately

Commit:
update sklnau8825adi (Caroline) audio support for 4.16.18

Audio now works properly under 4.16.13 and 4.16.18

Commit:
update debian/changelog

reynhout commented:
@bcarnes Awesome, this looks great. This means we should be able to include the same kernel for all platforms in 3.0, and jack detection is a great addition.

I'm curious about the `alsaucm` setup -- the delay was previously required for stable audio, but perhaps that's fixed in newer kernels. I haven't tested without the delay since 4.14.

Did you notice whether ACPI event `jack/microphone` is sent also, if the headphone plug has a mic? We've seen mixed results on other platforms (e.g. Braswell), but haven't tested Skylake.

Do you have any other changes queued up, or is this ready to package up for broader testing?

bcarnes commented:
> I'm curious about the alsaucm setup -- the delay was previously required for stable audio, but perhaps that's fixed in newer kernels.

Not sure it's particularly kernel or driver specific... You may have found the delay was needed due to a timing/race hack that was a result of alsa-restore(global), gallium-init-skylake(global) and then pulseaudio(local-user) all configuring some non-default essentials on the sound card.  You would've had a poisoned pulseaudio tdb cache of the bad default settings that was overriding your gallium-init-skylake settings unless you slept for a while.  I also found this confusing before I noticed the interactions.

For example, I found even with 4.16.18 (which has different timing than 4.16.13, perhaps due to the SD card bug), changing it to, say, "sleep 25" made things work again.  But better to remove the delay, set things immediately, and remove the poisoned pulseaudio saved settings.

Improving the ucm/*.conf files a bit further, and we can probably get rid of gallium-init-skylake completely.  And change the "n/a + n/a" to "Speaker + Headphones", etc.

In theory, we can even move from acpi+alsa jack detection to pulseaudio native jack detection, but a quick test of that looks like pulseaudio is a bit finicky with the current config, and I found acpi+alsa port swapping good enough for my uses.

Notice that manually moving from "n/a" (speaker) to "n/a" (headphones) in the GUI audio control (which uses pulseaudio) wedges pulseaudio permanently (or until the next pasuspend), as does going from headphones back to speaker.  Changing the port via acpi+alsa w/out pulseaudio involvement avoids this, but is dodging the fact that the alsa+pulseaudio configs are not right.

> Did you notice whether ACPI event jack/microphone is sent also, if the headphone plug has a mic?

Yes, we get acpi notifications for a TRRS microphone.  I think the current alsa&pulseaudio config will need a bit more tweaking before that'd just work, but if the microphone works when selected manually, we can acpi trigger it automatically.

> Do you have any other changes queued up, or is this ready to package up for broader testing?

I take it 3.0 is not being released imminently?  Why don't we leave this PR un-merged for developers to play with, and not merge/push it out more broadly until I take a crack at retiring gallium-init-skylake completely, and perhaps get the Mic working too.

For any developer playing with this (or other chromebook's sound settings):  I found the best "clean slate" way to test things short of a wipe+reinstall is:
```
  telinit 1
  rm -f /var/lib/alsa/*  # asound.state, used by alsa-restore
  rm -rf /home/*/.config/pulse # pulseaudio poisoned device tdb file
  reboot
```

Otherwise you'll be testing a confusing mishmash of: card reset state, alsa/ucm defaults, alsa saved state, pulseaudio card reset state, and pulseaudio saved state...! (and gallium-init-skylake reaching in to change things immediately or after N seconds, too)

Commit:
remove galliumos-init/skylake completely

use alsa's sklnau8825adi.conf SectionDefaults for init
remove bulk of init settings from alsa HiFi verb EnableSequence

Commit:
update debian/postinit

remove gallium-skylake/gallium-init-skylake service references

Commit:
rename "n/a" x 2 to "Speaker" and "Headphone"

Commit:
finish caroline alsa/pulseaudio support

  * jack events to use pulseaudio directly, rather than alsa
  * use alsa-ucm/pulseaudio profiles w/ single source/sink,
    to avoid multiple ports and pulseaudio freezes when switching
    between ports while sinks/sources are active.
  * Add microphone (built-in, and headset mic) support.

bcarnes commented:
I think this is ready to package up for broader testing...

doozan commented:
This works on Sentry hardware with a global s/sklnau8825adi/sklnau8825max on the patch set.  Are we going to need separate galliumos-skylake-caroline and galliumos-skylake-sentry packages, or is there a good way to combine everything?

MrChromebox commented:
@doozan we'll be handling all Skylake devices with the ADI codec identically. Sentry is a bit of a pain since some models have the ADI codec and some have the Maxim codec

Commit:
Add support for sklnau8825max devices (Sentry)

Commit:
relax acpi->PA script to work for both sklnau8825(adi|max)

bcarnes commented:
@doozan @MrChromebox
Easy enough to support both.  The alsa/ucm/* stuff is selected automatically by udev.  The acpi jack detection was intentionally restricting itself to just sklnau8825adi to not trigger on other systems that may have working jack detection natively(?).

Blindly added the bits needed for sklnau8825max without breaking sklnau8825adi.  See if this works for you, @doozan ...

MrChromebox commented on usr/bin/galliumos-init-skylake:
```diff
@@ -6,8 +6,7 @@ echo "configuring for $model"
 case "$model" in
   CAROLINE|CHELL)
```
needs to be expanded to cover all SKL CrOS devices

bcarnes replied:
This is in the gallium-skylake package.

I don't know how many other SKL CrOS devices there are, but according to:
  https://github.com/GalliumOS/galliumos-distro/issues/274
  https://github.com/GalliumOS/galliumos-distro/issues/379

on skylake only Caroline/skl*adi worked at all well before this PR, and this is a distinct improvement on what pre-existed.

Sounds like if it makes things better for Caroline (and Chell?, and now Sentry?), and doesn't make anything worse for any other models, we should merge this PR and eventually confirm success on other models or extend it to cover them.

This code is moving away from worrying about detecting "models" to just worrying about which chipsets were detected by udev - which sounds like it helps out on models like Sentry w/ a mix of adi & max.

I'll leave it to you GaOS developers who have access to these other devices to decide what to do next...

bcarnes replied:
Ah, @MrChromebox - note that switching on just Caroline and Chell was the pre-existing code before I got here. This commit is just fixing/removing the "sleep 5" hack. And by the end of this string of commits, I've removed this switching on model (and this entire galliumos-init-skylake script) completely...

reynhout commented:
@bcarnes Awesome. I will merge this and generate a pkg which I can test on CAROLINE and SENTRY/max98357.

reynhout commented:
@bcarnes Package tested successfully on CAROLINE/UEFI, and I believe @MrChromebox also had success on CHELL/UEFI.

Minor quirks on CAROLINE:

* On first install, audio was "spitty/clipped" (tested with `speaker-test -t sine` -- default 440Hz tone was not audible, just rapid popping which might be an artifact of the waveform). I dropped the ALSA speaker mixers down to 39% (from 40%), and the issue was resolved. I don't think I had previously changed the mixer settings from default install, will confirm. Headphone mixers were also at 40%, no issue there. Did not try goosing them to see if I could repro.
* Headphone switching works perfectly, but the first ~1/4 second of audio to the headphones is at max volume.

Not sure if these are ordinary ALSA/PA quirks and easily adjusted, or if we'd need to get hacky.

No luck on SENTRY/UEFI/max98357, but it might work on SENTRY/ssm4567.

The pkg is currently downloadable at https://galliumos/tmp/pkgs (version 2.9903), and I think it's ready to release to the `devel` and `testing` pkg repos (which are enabled by all 3.0beta users). Sound good? Thanks!

MrChromebox commented:
another issue is that the Speaker output is removed when the headphones are plugged in, so one has to unplug the headphone jack to use the speakers again. Not the end of the world, but seems like there should be a better way to handle that

bcarnes commented:
@MrChromebox @reynhout Good to hear it works for you both on Caroline & Chell

re: quirks:

* first install clipped until first volume adjustment
-> Never saw this myself, but should be addressed by more complete reset on first install found in #4 

* headphone switching causes brief audio boost to headphones
-> doesn't repro for me, I can set headphones @ say 30% and speakers @ 70%, and toggle back and forth without any 100% headphone burst.  Try it after #4, and see if it still happens for you now that we've reset *everything* we can to a known state.

* Sentry/skl*max
-> looks like @doozan forgot to mention he had a different kernel and a different dfw_sst.bin, looks like that is coming along now on #2

* speaker volume not present when headphones are in
-> think that's the standard pulseaudio behavior.  That's how it works on another (non-chromebook) laptop running ubuntu and a vanilla realtek HDA card.  There are other pulseaudio advanced rweaks to expose the non-active profile's volume, etc. but this is the right "out-of-the-box" behavior.

MrChromebox commented:
> speaker volume not present when headphones are in
> -> think that's the standard pulseaudio behavior. That's how it works on another (non-chromebook) laptop running ubuntu and a vanilla realtek HDA card. There are other pulseaudio advanced rweaks to expose the non-active profile's volume, etc. but this is the right "out-of-the-box" behavior.

FWIW, on another laptop I have here with a standard Realtek ALC256 codec, I loaded up GaOS 3.0, and it normally lists both 'Speakers' and 'Headphones (unplugged),' and when I plug in the headphones the status changed to 'Plugged' and it auto switches to them.
